[DO NOT SQUASH] MLIR#807: add tosa.cast to fusable elementwise ops in tosa-partition

sjw36 opened the pull request:
MLIR#812: fixed bug in tosa-to-linalg-named for quantized bias add
* added test
    
    https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/812

MLIR#813: fuse all subsequent transpose/reshape ops in tosa-partition
* also make tosa-partition run bottom-up (should use a pass instead)
* changed tosa-partition=trailing-only to default true
* updated tosa-partition tests
MLIR#807: add tosa.cast to fusable elementwise ops in tosa-partition
    
    https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/807
    https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/813

* fixed rock.transform_map for empty lowerBounds
* updated test for quantization kernel in partitioning

manupak commented on mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td:
```diff
@@ -192,7 +192,7 @@ def Rock_TransformMapAttr : Rock_Attr<"TransformMap"> {
 
     let assemblyFormat = [{
-        `<` $map `by` ` ` `[` $ops `]` `bounds` `=` ($upperBounds^) : (`[` `]`)? `->` $lowerBounds `>`
+        `<` $map `by` ` ` `[` $ops `]` `bounds` `=` ($upperBounds^) : (`[` `]`)? `->` ($lowerBounds^) : (`[` `]`)? `>`
```
Why this change ?

sjw36 replied:
I found with new partitioning where an empty lower bounds happens. It's like a memref.collapse_shape : memref<1x1x1xf32> into memref<f32>

manupak replied:
Ack!

manupak commented on external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaPartition.cpp:
```diff
@@ -562,7 +567,7 @@ void TosaPartitionPass::runOnOperation() {
       // Grab a useful set of leading ops, like we do for trailing.
       SetVector<Operation *> leadingOps;
-      traceInputs(anchorOp, leadingOps, inputNodes);
+      traceInputs(anchorOp, anchorOp, leadingOps, inputNodes);
```
Why anchor op as ignore op here ?

sjw36 replied:
Since this routine is now used below for walking up inputs to trailing elementwise ops, the ignoreOp was added so we don't recurse back up the anchor op path. But here it is mostly a placeholder (vs nullptr) since we want to walk back every input of the anchor op.

manupak replied:
Ack!

manupak reviewed (commented):
Changes broadly looks good!
just two questions..

manupak reviewed (approved):
LGTM!

pcf000 reviewed (approved):
A few things I did differently in my generic-partition branch;  we'll harmonise later.

Commit:
MLIR#812: fixed bug in tosa-to-linalg-named for quantized bias add
    * added test

    https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/812

Commit:
MLIR#813: fuse all subsequent transpose/reshape ops in tosa-partition
      * also make tosa-partition run bottom-up (should use a pass instead)
      * changed tosa-partition=trailing-only to default true
      * updated tosa-partition tests
MLIR#807: add tosa.cast to fusable elementwise ops in tosa-partition

    https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/807
    https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/813

Commit:
* fixed rock.transform_map for empty lowerBounds
    * updated test for quantization kernel in partitioning
